commons/src/main/kotlin/com/github/quillraven/commons/audio/QueueAudioService.kt

```diff
@@ -0,0 +1,190 @@
+ */
+private class SoundRequest : Pool.Poolable {
+  var filePath: String = ""
+  var volume = 1f
+
+  override fun reset() {
+    filePath = ""
+    volume = 1f
+  }
+}
+
+/**
+ * Implementation of [AudioService] to play one [Music] track at a time and to queue [Sound] playback
+ * to avoid multiple [Sound] instances of the same sound within a single frame.
+ *
+ * Requires an [assetStorage] to load and unload the [Sound] and [Music] resources.
+ *
+ * Use [maxSimultaneousSounds] to define the maximum number of sounds that can play concurrently within one frame.
+ *
+ * Use [maxCachedSounds] to define the size of the sound cache. [Sound] instances are cached internally for
+ * faster access of future calls. If the [maxCachedSounds] is reached then the cache gets cleared.
+ */
+class QueueAudioService(
+  private val assetStorage: AssetStorage,
+  private val maxSimultaneousSounds: Int = 16,
+  private val maxCachedSounds: Int = 50
+) : AudioService {
+
+  private val soundRequests = ObjectMap<String, SoundRequest>(maxSimultaneousSounds)
+  private val soundCache = ObjectMap<String, Sound>(maxCachedSounds)
+  private var currentMusicFilePath: String = ""
+  private var currentMusic: Music? = null
+
+  /**
+   * Adds a [SoundRequest] to the request queue for the [soundFilePath] and [volume].
+   * If the [maxSimultaneousSounds] is reached for the current frame then this request is ignored.
+   * Making a call to [update] will play all pending requests.
+   */
+  override fun playSound(soundFilePath: String, volume: Float) {
+    if (soundRequests.size >= maxSimultaneousSounds) {
+      LOG.info { "Reached maximum of '$maxSimultaneousSounds' simultaneous sounds. Request is ignored" }
+      return
+    }
+
+    val request = soundRequests[soundFilePath]
+    if (request != null) {
+      // sound already queued -> set volume to maximum of both requests
+      request.volume = max(request.volume, volume).coerceIn(0f, 1f)
+    } else {
+      // sound not queued yet
+      soundRequests[soundFilePath] = SOUND_REQUEST_POOL.obtain().apply {
+        this.volume = volume
+        this.filePath = soundFilePath
+      }
+    }
+  }
+
+  /**
+   * Loads and plays a new [Music] instance for [musicFilePath] with the given [volume].
+   * If [loop] is true then the music will start again when finished.
+   *
+   * If there is already a music playing then it gets stopped and unloaded from the [assetStorage].
+   */
+  override fun playMusic(musicFilePath: String, volume: Float, loop: Boolean) {
+    if (currentMusicFilePath == musicFilePath) {
+      // trying to play the same music again -> ignore it
+      LOG.debug { "Music '$musicFilePath' is already playing" }
+      return
+    }
+
+    val oldMusic = currentMusic
+    if (oldMusic != null) {
+      // stop and unload current music
+      LOG.debug { "Stop music '$currentMusicFilePath'" }
+      oldMusic.stop()
+      KtxAsync.launch {
+        assetStorage.unload<Music>(currentMusicFilePath)
```

Comment:
It's not necessary to unload the music after it is finished. This will cause noticeable delays each time you want to change the theme. Let the engine users pre-load and manage which files should be loaded or not at any given time.